crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
 /// That means that when you are constructing a formula, you might need to create distinct atoms
 /// for a particular [`Constraint`], if your formula needs to consider how a particular runtime
 /// property might be different at different points in the execution of the program.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) struct VisibilityConstraint<'db>(VisibilityConstraintInner<'db>);
-
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) enum VisibilityConstraintInner<'db> {
-    AlwaysTrue,
-    AlwaysFalse,
-    Ambiguous,
-    VisibleIf(Constraint<'db>, u8),
-    VisibleIfNot(ScopedVisibilityConstraintId),
-    KleeneAnd(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
-    KleeneOr(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+///
+/// Visibility constraints are normalized, so equivalent constraints are guaranteed to have equal
+/// IDs.
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+pub(crate) struct ScopedVisibilityConstraintId(u32);
+
+impl std::fmt::Debug for ScopedVisibilityConstraintId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
+        match *self {
+            ALWAYS_TRUE => f.field(&format_args!("AlwaysTrue")).finish(),
+            AMBIGUOUS => f.field(&format_args!("Ambiguous")).finish(),
+            ALWAYS_FALSE => f.field(&format_args!("AlwaysFalse")).finish(),
+            _ => f.field(&self.0).finish(),
+        }
```

Comment:
Nit: Makes it clearer what's different/shared between the branches:

```suggestion
        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
        match *self {
            ALWAYS_TRUE => f.field(&"AlwaysTrue"),
            AMBIGUOUS => f.field(&"Ambiguous"),
            ALWAYS_FALSE => f.field(&"AlwaysFalse"),
            _ => f.field(&self.0),
        }
        
        f.finish()
```